File: output_repair/claude-opus/sample_3/repair_3/c/c/Struct/deposit.c
#include "stdlib.h"

struct account
{
    int balance;
};

/*@ predicate account(struct account *a; int b) = malloc_block_account(a) &*& a->balance |-> b; @*/

struct account *create_account()
//@ requires true;
//@ ensures account(result, 0);
{
    struct account *myAccount = malloc(sizeof(struct account));
    if (myAccount == 0)
    {
        abort();
    }
    myAccount->balance = 0;
    return myAccount;
}

void account_deposit(struct account *myAccount, int amount)
//@ requires account(myAccount, ?b) &*& INT_MIN - b <= amount &*& amount <= INT_MAX - b;
//@ ensures account(myAccount, b + amount);
{
    myAccount->balance += amount;
}

int account_get_balance(struct account *myAccount)
//@ requires account(myAccount, ?b);
//@ ensures account(myAccount, b) &*& result == b;
{
    int result = myAccount->balance;
    return result;
}

void account_set_balance(struct account *myAccount, int newBalance)
//@ requires account(myAccount, _);
//@ ensures account(myAccount, newBalance);
{
    myAccount->balance = newBalance;
}

void account_dispose(struct account *myAccount)
//@ requires account(myAccount, _);
//@ ensures true;
{
    free(myAccount);
}

int main()
//@ requires true;
//@ ensures true;
{
    struct account *myAccount = create_account();
    account_set_balance(myAccount, 5);
    account_deposit(myAccount, 10);
    int b = account_get_balance(myAccount);
    assert(b == 15);
    account_dispose(myAccount);
    return 0;
}

File: output_log/claude-opus/sample_2/rust/qwen3/rust/purely_unsafe/solutions/deposit.rs
use std::alloc::{alloc, dealloc, handle_alloc_error, Layout};

struct Account {
    balance: i32,
}

#[predicate]
pub account_pred(p: *mut Account) =
    p != 0 &*&
    malloc_block_Account(p, 1) &*&
    p->Account(balance: ?b);

#[predicate]
pub acc_balance(p: *mut i32, b: i32) = p->b;

impl Account {
    unsafe fn create() -> *mut Account {
        #[requires(true)]
        #[ensures(account_pred(result))]
        {
            let my_account = alloc(Layout::new::<Account>()) as *mut Account;
            if my_account.is_null() {
                handle_alloc_error(Layout::new::<Account>());
            }
            // after allocation we have malloc_block but no initialization yet
            #[requires(my_account != 0 &*& malloc_block_Account(my_account, 1))]
            #[ensures(account_pred(my_account) &*& *my_account == Account { balance: 0 })]
            {
                (*my_account).balance = 0;
            }
            my_account
        }
    }

    unsafe fn get_balance(my_account: *mut Account) -> i32 {
        #[requires(account_pred(my_account))]
        #[ensures(account_pred(my_account) &*& result == (*my_account).balance)]
        {
            (*my_account).balance
        }
    }

    unsafe fn set_balance(my_account: *mut Account, new_balance: i32) {
        #[requires(account_pred(my_account))]
        #[ensures(account_pred(my_account) &*& (*my_account).balance == new_balance)]
        {
            (*my_account).balance = new_balance;
        }
    }

    unsafe fn deposit(my_account: *mut Account, amount: i32) {
        #[requires(account_pred(my_account))]
        #[ensures(account_pred(my_account) &*& (*my_account).balance == old((*my_account).balance) + amount)]
        {
            (*my_account).balance += amount;
        }
    }

    unsafe fn dispose(my_account: *mut Account) {
        #[requires(account_pred(my_account))]
        #[ensures(true)]
        {
            dealloc(my_account as *mut u8, Layout::new::<Account>());
        }
    }
}

fn main() {
    unsafe {
        let my_account = Account::create();
        Account::set_balance(my_account, 5);
        Account::deposit(my_account, 10);
        let b = Account::get_balance(my_account);
        std::hint::assert_unchecked(b == 15);
        Account::dispose(my_account);
    }
}